### PR TITLE
Added GenericSchemeMultiCall

### PR DESCRIPTION
**Pull request by nicoelzer**
- Added GenericSchemeMultiCall

### CODE REVIEW

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+ * @title GenericScheme.
+ * @dev  A scheme for proposing and executing calls to multiple arbitrary function
+ * on whitelisted contract on behalf of the organization avatar.
+ */
+contract GenericSchemeMulticall is VotingMachineCallbacks, ProposalExecuteInterface {
+    event NewMultiCallProposal(
+        address indexed _avatar,
+        bytes32 indexed _proposalId,
+        bytes[]   _callData,
+        uint[] _value,
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:18]: uint256[]

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    // Details of a voting proposal:
+    struct MultiCallProposal {
+        address[] contractToCall;
+        bytes[] callData;
+        uint256[] value;
+        bool exist;
+        bool passed;
+    }
+
+    mapping(bytes32=>MultiCallProposal) public organizationProposals;
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:52]: rename organizationProposals -> proposals

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    * @dev propose to call on behalf of the _avatar
+    *      The function trigger NewCallProposal event
+    * @param _contractToCall the contract to be called 
+    * @param _callData - The abi encode data for the call
+    * @param _value value(ETH) to transfer with the call
+    * @param _descriptionHash proposal description hash
+    * @return an id which represents the proposal
+    */
+
+    function proposeCall(address[] memory _contractToCall, bytes[] memory _callData, uint256[] memory _value, string memory _descriptionHash)
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:140]: `proposeCalls` ?

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        delete organizationProposals[_proposalId];
+        emit ProposalDeleted(address(avatar), _proposalId);
+        emit ProposalExecuted(address(avatar), _proposalId);
+
+    }
+
+    /**
+    * @dev propose to call on behalf of the _avatar
+    *      The function trigger NewCallProposal event
+    * @param _contractToCall the contract to be called
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:133]: description is wrong..

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        require(proposal.passed, "proposal must passed by voting machine");
+        proposal.exist = false;
+        bytes memory genericCallReturnValue;
+        bool success;
+        Controller controller = Controller(avatar.owner());
+
+        for (uint i = 0; i < proposal.contractToCall.length; i ++) {
+            (success, genericCallReturnValue) =
+            controller.genericCall(proposal.contractToCall[i], proposal.callData[i], avatar, proposal.value[i]);
+            require(success, "Proposal call failed");
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:120]: so if one of the calls fails all their is no execution .. maybe add a comment re that

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    /**
+     * @dev initialize
+     * @param _avatar the avatar to mint reputation from
+     * @param _votingMachine the voting machines address to
+     * @param _voteParams voting machine parameters.
+     */
+    function initialize(
+        Avatar _avatar,
+        IntVoteInterface _votingMachine,
+        bytes32 _voteParams
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:67]: how about adding here a white listed targets contracts? for security reasons.

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 3 above)
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:140]: _contractsToCall

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    * @param _contractToCall the contract to be called 
+    * @param _callData - The abi encode data for the call
+    * @param _value value(ETH) to transfer with the call
+    * @param _descriptionHash proposal description hash
+    * @return an id which represents the proposal
+    */
+
+    function proposeCall(address[] memory _contractToCall, bytes[] memory _callData, uint256[] memory _value, string memory _descriptionHash)
+    public
+    returns(bytes32)
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:142]: `returns(bytes32 proposalId)`

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+            exist: true,
+            passed: false
+        });
+        proposalsInfo[address(votingMachine)][proposalId] = ProposalInfo({
+            blockNumber:block.number,
+            avatar:avatar
+        });
+
+        emit NewMultiCallProposal(address(avatar), proposalId, _callData, _value, _descriptionHash, _contractToCall);
+        return proposalId;
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:163]: can be removed

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 9]
```diff
@@ -0,0 +1,167 @@
+pragma solidity 0.5.17;
+pragma experimental ABIEncoderV2;
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:2]: why this is needed ?

**Review by orenyodfat — CHANGES_REQUESTED** [review]

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 9 above)
Comment (nicoelzer) [reply contracts/schemes/GenericSchemeMultiCall.sol:2]: @orenyodfat bytes[]  _callData is only supported with ABIEncoderV2 in 0.5

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 8 above)
Comment (nicoelzer) [reply contracts/schemes/GenericSchemeMultiCall.sol:163]: Why removing it? We would still want to emit the event and return proposalId after creating the Proposal

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 9 above)
Comment (orenyodfat) [reply contracts/schemes/GenericSchemeMultiCall.sol:2]: ok. the other alternative is to pass in bytes _callData which is concatenation of call data of all calls .
and provide additional lens array to parse it.
something like https://github.com/daostack/arc/blob/master-2/contracts/utils/DAOFactory.sol#L235 
this will eliminate the need to use the `experimental` pragma which is not experimental.

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 6 above)
Comment (nicoelzer) [reply contracts/schemes/GenericSchemeMultiCall.sol:67]: @orenyodfat as we would need to call allowance() on some tokens we don't know at initialization we won't be able to whitelist specific contract addresses. Thinking about a possible solution for restriction right now.

Comment (orenyodfat) [reply contracts/schemes/GenericSchemeMultiCall.sol:67]: losing a bit from generality , one option is to add explicit none restricted `approval` token calls as part of a proposal .
using https://github.com/daostack/arc/blob/master/contracts/controller/Controller.sol#L460

Comment (nicoelzer) [reply contracts/schemes/GenericSchemeMultiCall.sol:67]: We could still keep the generality if we would include an encoded call within the proposalCalls function that would execute externalTokenApproval() on the controller before sending funds. In that way we could still use a contract whitelist as we would always use the controller (known address) for approvals – correct?

Comment (orenyodfat) [reply contracts/schemes/GenericSchemeMultiCall.sol:67]: > @orenyodfat we could still keep the generality if we would include an encoded call within the proposalCalls function that would execute externalTokenApproval() on the controller before sending funds. In that way we could still use a contract whitelist as we would always use the controller (known address) for approvals – correct?

no. because the avatar is not allowed to call controller . it is not registered scheme

Comment (orenyodfat) [reply contracts/schemes/GenericSchemeMultiCall.sol:67]: > @orenyodfat we could still keep the generality if we would include an encoded call within the proposalCalls function that would execute externalTokenApproval() on the controller before sending funds. In that way we could still use a contract whitelist as we would always use the controller (known address) for approvals – correct?

no . because the avatar is not allowed to call controller .(it is not a register scheme)

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 9 above)
Comment (nicoelzer) [reply contracts/schemes/GenericSchemeMultiCall.sol:2]: @orenyodfat the DAOfactory is also using ABIEncoderV2, so I guess that shouldn't be an issue

Comment (orenyodfat) [reply contracts/schemes/GenericSchemeMultiCall.sol:2]: DAOFactory use solidity 6.12

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+            uint256 valueToSpend
+           ) =
+           /* solhint-disable */
+           abi.decode(
+             proposal.callData[i],
+             (IERC20, address, uint256)
+           );
+           controller.externalTokenApproval(extToken,spender,valueToSpend,avatar);
+         } else {
+           require(contractWhitelist[proposal.contractsToCall[i]], "contractToCall is not whitelisted");
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:143]: this should be check at the proposal function

File: test/genericscheme.js [hunk 11]
```diff
@@ -232,7 +232,7 @@ contract('GenericScheme', function(accounts) {
         });
       });
 
-      it("Wallet - execute proposeVote -positive decision - check action - with GenesisProtocol", async function() {
+      it.only("Wallet - execute proposeVote -positive decision - check action - with GenesisProtocol", async function() {
```
Comment (orenyodfat) [inline test/genericscheme.js:235]: is this related to this pr ?

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+    * @param _value value(ETH) to transfer with the calls
+    * @param _descriptionHash proposal description hash
+    * @return an id which represents the proposal
+    */
+
+    function proposeCalls(address[] memory _contractsToCall, bytes[] memory _callData, uint256[] memory _value, string memory _descriptionHash)
+    public
+    returns(bytes32 proposalId)
+    {
+        require(
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:170]: is there a test covering this one ?

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 10 above)
Comment (orenyodfat) [reply contracts/schemes/GenericSchemeMultiCall.sol:143]: any test cover proposing none white list contracts ?

File: contracts/schemes/GenericSchemeMultiCall.sol [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+           (IERC20 extToken,
+            address spender,
+            uint256 valueToSpend
+           ) =
+           /* solhint-disable */
+           abi.decode(
+             proposal.callData[i],
+             (IERC20, address, uint256)
+           );
+           controller.externalTokenApproval(extToken,spender,valueToSpend,avatar);
```
Comment (orenyodfat) [inline contracts/schemes/GenericSchemeMultiCall.sol:141]: any test cover this case ?

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 10 above)
Comment (nicoelzer) [reply contracts/schemes/GenericSchemeMultiCall.sol:143]: > any test cover proposing none white list contracts ?

it("execute proposeVote -positive decision - not whitelisted contract")

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 12 above)
Comment (nicoelzer) [reply contracts/schemes/GenericSchemeMultiCall.sol:170]: > is there a test covering this one ?

@orenyodfat it("proposeCall log - with invalid array - reverts")

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 13 above)
Comment (nicoelzer) [reply contracts/schemes/GenericSchemeMultiCall.sol:141]: it("execute proposeVote with multiple calls with votingMachine -positive decision")

File: contracts/schemes/GenericSchemeMultiCall.sol (hunk 6 above)
Comment (AugustoL) [reply contracts/schemes/GenericSchemeMultiCall.sol]: @orenyodfat the avatar wont call the controller it would called an ERC20 contract and execute the approve function.

Comment (AugustoL) [reply contracts/schemes/GenericSchemeMultiCall.sol]: But taht will require the ERC20 token to be listed in the whitelisted contracts, right? this way the controller can call the approve function in any token contract, right? or am I missing something?

Comment (orenyodfat) [reply contracts/schemes/GenericSchemeMultiCall.sol]: @nicoelzer  does any how the plan is to allow the avatar to transfer any erc20 token its own?. this might be the same as approval ... if this is the case we might do not need this special approval staff.
If the case is that only approval is needed for any erc20 so we can avoid white listing the erc20 and have the solution as it is now .
